# Analyzing Ebay Car Sales Data

Goal: Clean the data to analyze the included used car listings

Data source: https://data.world/data-society/used-cars-data

## 1. Importing the data

In [147]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [148]:
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

## 2. Observing the data

In [149]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [150]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Observations:

* The dataset contains 20 columns, most of which are strings.

* Some columns have null values, but none have more than 20% null values.

* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

* The 'dateCrawled' column is the date when this ad was first crawled and the 'name' column shows the name of the car.

* The content of the data is in german, as it can be found from the values in the 'notRepairedDamage' column.

## 3. Renaming the columns

In [151]:
cols = autos.columns.copy()

# The original column names
print(cols)

alist = []
for c in cols:
    if c == 'yearOfRegistration':
        c = 'registration_year'
    elif c == 'monthOfRegistration':
        c = 'registration_month'
    elif c == 'notRepairedDamage':
        c = 'unrepaired_damage'
    elif c == 'dateCreated':
        c = 'ad_created'
    elif c == 'dateCrawled':
        c = 'date_crawled'
    elif c == 'offerType':
        c = 'offer_type'
    elif c == 'vehicleType':
        c = 'vehicle_type'
    elif c == 'gearbox':
        c = 'gear_box'
    elif c == 'powerPS':
        c = 'power_PS'
    elif c == 'fuelType':
        c = 'fuel_type'
    elif c == 'nrOfPictures':
        c = 'nr_of_pictures'
    elif c == 'postalCode':
        c = 'postal_code'
    elif c == 'lastSeen':
        c = 'last_seen'
    else:
        c = c
    alist.append(c)
                 
# The new column names
print(alist)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')
['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']


In [152]:
autos.columns = alist
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In order to avoid confusion with DataFrame methods and properties, camel case column names were changed to snake case. It is also easier to extend to slices and indexing.


Next, we explore the dataset to further clean the data.

## 4. Dropping columns with insignificant values

In [153]:
# Explore the entire dataset
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Let's look for text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

In [154]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [155]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [156]:
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [157]:
autos = autos.drop(['seller', 'offer_type', 'nr_of_pictures'], axis=1)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year   gear_box  power_PS  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## 5. Converting text columns to numeric type

Numeric data stored as text('price' and 'odometer' columns) will also be converted to integers.

In [158]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos.rename({'price':'price_dollar'}, axis=1, inplace=True)

In [159]:
autos['price_dollar'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_dollar, dtype: int64

In [160]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [161]:
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## 6. Getting rid of outliers

Let's continue analyzing the 'odometer_km' and 'price_dollar' columns

In [162]:
autos['price_dollar'].unique().shape

(2357,)

In [163]:
autos['price_dollar'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64

In [164]:
autos['price_dollar'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_dollar, dtype: int64

In [165]:
autos['price_dollar'].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price_dollar, dtype: int64

We can find that the maximum value '99999999' is an outlier, so we get rid of this row by changing the range.

In [166]:
autos = autos[autos['price_dollar'].between(0,27322222)]
print(autos['price_dollar'].value_counts().sort_index(ascending=True))

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
145000         1
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111      

In [167]:
autos['odometer_km'].unique().shape

(13,)

In [168]:
autos['odometer_km'].describe()

count     49999.000000
mean     125732.214644
std       40042.465064
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [169]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32423
Name: odometer_km, dtype: int64

We can't find an outlier for the 'odometer_km' column, so we leave this column as it is.

## 7. Exploring date columns


Let's now move on to the date columns and understand the date range the data covers.


'date_crawled', 'last_seen', and 'ad_created' columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. On the other hand, 'registration_month' and 'registration_year' columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.


Let's first see how the values in these comlumns are formatted.

In [170]:
autos[['date_crawled','ad_created','last_seen','registration_month','registration_year']][0:5]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

Next, let's calculate the distribution of values in the 'date_crawled', 'ad_created', and 'last_seen' columns as percentages.

In [171]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2016-03-05    0.025381
2016-03-06    0.013940
2016-03-07    0.035961
2016-03-08    0.033301
2016-03-09    0.033221
2016-03-10    0.032121
2016-03-11    0.032481
2016-03-12    0.036781
2016-03-13    0.015560
2016-03-14    0.036621
2016-03-15    0.033981
2016-03-16    0.029501
2016-03-17    0.031521
2016-03-18    0.013060
2016-03-19    0.034901
2016-03-20    0.037821
2016-03-21    0.037521
2016-03-22    0.032921
2016-03-23    0.032381
2016-03-24    0.029101
2016-03-25    0.031741
2016-03-26    0.032481
2016-03-27    0.031041
2016-03-28    0.034841
2016-03-29    0.034181
2016-03-30    0.033621
2016-03-31    0.031921
2016-04-01    0.033801
2016-04-02    0.035401
2016-04-03    0.038681
2016-04-04    0.036521
2016-04-05    0.013100
2016-04-06    0.003180
2016-04-07    0.001420
Name: date_crawled, dtype: float64


In [172]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033241
2016-03-10    0.031861
2016-03-11    0.032781
2016-03-12    0.036621
2016-03-13    0.016920
2016-03-14    0.035221
2016-03-15    0.033741
2016-03-16    0.030001
2016-03-17    0.031201
2016-03-18    0.013720
2016-03-19    0.033841
2016-03-20    0.037861
2016-03-21 

In [173]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2016-03-05    0.001080
2016-03-06    0.004420
2016-03-07    0.005360
2016-03-08    0.007600
2016-03-09    0.009860
2016-03-10    0.010760
2016-03-11    0.012520
2016-03-12    0.023820
2016-03-13    0.008980
2016-03-14    0.012800
2016-03-15    0.015880
2016-03-16    0.016440
2016-03-17    0.027921
2016-03-18    0.007420
2016-03-19    0.015740
2016-03-20    0.020700
2016-03-21    0.020740
2016-03-22    0.021580
2016-03-23    0.018580
2016-03-24    0.019560
2016-03-25    0.019200
2016-03-26    0.016960
2016-03-27    0.016020
2016-03-28    0.020860
2016-03-29    0.022340
2016-03-30    0.024840
2016-03-31    0.023840
2016-04-01    0.023100
2016-04-02    0.024900
2016-04-03    0.025361
2016-04-04    0.024620
2016-04-05    0.124282
2016-04-06    0.220984
2016-04-07    0.130923
Name: last_seen, dtype: float64


## 8. Cleaning date columns

For all three columns, we can find that maximum values were reached on April 3rd. Now, let's observe the 'registration_month' and 'registration_year' column.

In [174]:
autos['registration_month'].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  0,  9, 11,  5,  2,  1])

'0' month seems to be a problem. Let's observe how many data points have '0' month.

In [175]:
autos[autos['registration_month']==0].describe()

price_dollar  registration_year     power_PS    odometer_km  \
count  5.075000e+03        5075.000000  5075.000000    5075.000000   
mean   6.787435e+03        2012.730049    66.869754  127903.448276   
std    2.331412e+05         285.652252    84.319912   44561.748688   
min    0.000000e+00        1000.000000     0.000000    5000.000000   
25%    3.000000e+02        1997.000000     0.000000  125000.000000   
50%    9.000000e+02        2000.000000    60.000000  150000.000000   
75%    2.400000e+03        2006.000000   113.000000  150000.000000   
max    1.234568e+07        9999.000000  1998.000000  150000.000000   

       registration_month   postal_code  
count              5075.0   5075.000000  
mean                  0.0  48120.804335  
std                   0.0  25955.661438  
min                   0.0   1067.000000  
25%                   0.0  26906.000000  
50%                   0.0  46325.000000  
75%                   0.0  66613.000000  
max                   0.0  99986.000000

In [176]:
print("Percentage of '0' registration_month:{}".format(len(autos[autos['registration_month']==0])/len(autos)*100))

Percentage of '0' registration_month:10.150203004060081


Data points with '0' registration_month column takes over 5%, which is a substantial amount, so we will leave it as it is.

In [177]:
autos['registration_year'].describe()

count    49999.000000
mean      2005.073401
std        105.713866
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There seems to be some strange values in the 'registration_year' column(e.g. 1000, 9999), so let's observe it further.

In [178]:
autos['registration_year'].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 4500, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1959,
       9996, 9999, 6200, 1964, 1958, 1800, 1948, 1931, 1943, 9000, 1941,
       1962, 1927, 1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888,
       1954, 1938, 2800, 5911, 1500, 1953, 1951, 4800, 1001])

In [179]:
with pd.option_context('display.max_rows', 999):
    print(autos['registration_year'].value_counts())

2000    3354
2005    3015
1999    2999
2004    2737
2003    2727
2006    2708
2001    2703
2002    2533
1998    2453
2007    2304
2008    2231
2009    2098
1997    2028
2011    1634
2010    1597
2017    1453
1996    1444
2012    1323
2016    1316
1995    1313
2013     806
2014     666
1994     660
2018     492
1993     445
2015     399
1990     395
1992     391
1991     356
1989     181
1988     142
1985     105
1980      97
1986      76
1987      75
1983      53
1984      53
1978      47
1970      45
1982      43
1972      35
1979      35
1960      34
1981      31
1971      27
1967      27
1976      27
1968      26
1973      26
1974      24
1966      22
1977      22
1975      19
1969      19
1965      17
1964      12
1910       9
1963       9
1959       7
1961       6
1956       5
5000       4
9999       4
1958       4
1962       4
1937       4
2019       3
1950       3
1955       2
9000       2
1954       2
1800       2
1957       2
1941       2
1951       2
1934       2
4100       1

We will only consider values from 1960 to 2016 because 2016 is the date when this data was crawled, and it wouldn't make sense to think that data was crawled before 1960 when computer was barely available.

In [180]:
autos = autos.loc[autos['registration_year'].between(1960,2016)]
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_dollar   abtest vehicle_type  registration_year   gear_box  power_PS  \
0          5000  control          bus               2004    manuell       158   
1          8500  control    limousine               1997  automatik       286   
2          8990     test    limousine               2009    manuell       102   
3          4350  control   kleinwagen               2007  automatik        71   
4          1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_ty

## 9. Observing car sales amount by year

In [181]:
autos['registration_year'].value_counts(normalize=True) * 100

2000    6.991287
2005    6.284654
1999    6.251303
2004    5.705174
2003    5.684329
2006    5.644724
2001    5.634302
2002    5.279943
1998    5.113186
2007    4.802601
2008    4.650436
2009    4.373202
1997    4.227290
2011    3.406012
2010    3.328886
1996    3.009964
2012    2.757744
2016    2.743153
1995    2.736899
2013    1.680077
2014    1.388252
1994    1.375745
1993    0.927586
2015    0.831701
1990    0.823363
1992    0.815025
1991    0.742069
1989    0.377288
1988    0.295994
1985    0.218869
1980    0.202193
1986    0.158419
1987    0.156335
1984    0.110477
1983    0.110477
1978    0.097970
1970    0.093801
1982    0.089632
1972    0.072956
1979    0.072956
1960    0.070872
1981    0.064618
1971    0.056280
1967    0.056280
1976    0.056280
1973    0.054196
1968    0.054196
1974    0.050027
1966    0.045858
1977    0.045858
1975    0.039605
1969    0.039605
1965    0.035436
1964    0.025014
1963    0.018760
1961    0.012507
1962    0.008338
Name: registration_year, dtype:

From the above observation, we can find that the amount of car sales peaked from the late 1990's to ealry 2000's.

## 10. Observing car sales amount by brand

In [182]:
br_series = autos['brand'].value_counts(normalize=True) * 100
print(br_series)

volkswagen        21.232334
bmw               11.012215
opel              10.820444
mercedes_benz      9.528078
audi               8.648435
ford               6.974611
renault            4.737983
peugeot            2.955768
fiat               2.588902
seat               1.817651
skoda              1.602952
mazda              1.515404
nissan             1.511235
smart              1.392421
citroen            1.392421
toyota             1.248593
sonstige_autos     1.052654
hyundai            0.985951
volvo              0.923417
mini               0.865052
mitsubishi         0.815025
honda              0.785842
kia                0.710802
alfa_romeo         0.660775
porsche            0.608663
suzuki             0.591987
chevrolet          0.566974
chrysler           0.366865
daihatsu           0.256389
dacia              0.256389
jeep               0.225122
subaru             0.218869
land_rover         0.202193
saab               0.160504
jaguar             0.158419
trabant            0

## 11. Aggregating the data by top sales brands and mileage

Let's observe average price of each brand that were registered over 5%.

In [183]:
price_dict = {}
for i, r in br_series.iteritems():
    if r >= 5:
        selected_rows = autos[autos['brand']==i]
        selected_price_mean = selected_rows['price_dollar'].mean()
        price_dict[i] = selected_price_mean
print(price_dict)

{'bmw': 8334.045996592846, 'opel': 5253.836062415719, 'ford': 7233.840406455469, 'mercedes_benz': 8422.67818858018, 'audi': 9093.65003615329, 'volkswagen': 6513.250539956804}


From the aggregated data, we can find that the most expensive brands are BMW, Mercedes Benz, Audi, while Opel and Volkswagen are less expensive, and Ford is in between. For these top 6 brands, let's use aggregation to understand the average mileage for these cars and if there's any visible link with mean price.

In [184]:
mileage_dict = {}
for k, v in price_dict.items():
    selected_rows = autos[autos['brand']==k]
    selected_mileage_mean = selected_rows['odometer_km'].mean()
    mileage_dict[k] = selected_mileage_mean
print(mileage_dict)

{'bmw': 132458.8302101079, 'opel': 129258.33172799075, 'ford': 124176.62881052002, 'mercedes_benz': 130985.56114635748, 'audi': 129287.78018799711, 'volkswagen': 128731.10151187905}


In [185]:
s1 = pd.Series(price_dict)
s2 = pd.Series(mileage_dict)
df1 = pd.DataFrame(s1, columns=['mean_price'])
df2 = pd.DataFrame(s2, columns=['mean_mileage'])
df = df1.merge(df2, left_index=True, right_index=True)
df.round(2)

mean_price  mean_mileage
audi              9093.65     129287.78
bmw               8334.05     132458.83
ford              7233.84     124176.63
mercedes_benz     8422.68     130985.56
opel              5253.84     129258.33
volkswagen        6513.25     128731.10

As it can be found from the above result, there is no clear relationship between each brand's price and its mileage.